File: ghidra-scripts/ExtractShotPipeline.java
// Extracts likely shot-pipeline functions by matching interesting strings and
// demangled function names, then decompiles them to text files.

import ghidra.app.decompiler.DecompInterface;
import ghidra.app.decompiler.DecompileResults;
import ghidra.app.script.GhidraScript;
import ghidra.program.model.listing.Data;
import ghidra.program.model.listing.DataIterator;
import ghidra.program.model.listing.Function;
import ghidra.program.model.listing.FunctionIterator;
import ghidra.program.model.listing.Listing;
import ghidra.program.model.symbol.Reference;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.Map;
import java.util.Set;
import java.util.TreeMap;
import java.util.TreeSet;

public class ExtractShotPipeline extends GhidraScript {

	private static final String[] STRING_KEYWORDS = {
		"Wait.For.Shot",
		"waitForShot",
		"ExtractBallClub",
		"Extract.BallClub",
		"Ball.Build",
		"Club.Build",
		"BackSpinSideSpinInference",
		"Club.Detection",
		"Ball.Impact.On.Club",
		"loadCamera",
		"StartSingleFrameCamera",
		"StartBoardFrameCamera",
		"saveShot",
		"exportPostShot",
		"loadMLModel"
	};

	private static final String[] FUNCTION_KEYWORDS = {
		"waitForShot",
		"extractBallClub",
		"ballHasMoveEnough",
		"saveShot",
		"exportPostShot",
		"loadCamera",
		"loadMLModel",
		"setCameraCallback",
		"onNewShot",
		"spin",
		"club"
	};

	private static class Match {
		final Function function;
		final Set<String> reasons = new TreeSet<>();

		Match(Function function) {
			this.function = function;
		}
	}

	@Override
	protected void run() throws Exception {
		String[] args = getScriptArgs();
		if (args.length < 1) {
			throw new IllegalArgumentException("Expected output directory as first argument");
		}

		Path outputDir = Paths.get(args[0]);
		Files.createDirectories(outputDir);

		Map<String, Match> interesting = new TreeMap<>();

		Listing listing = currentProgram.getListing();
		DataIterator dataIterator = listing.getDefinedData(true);
		while (dataIterator.hasNext() && !monitor.isCancelled()) {
			Data data = dataIterator.next();
			if (!data.hasStringValue()) {
				continue;
			}

			Object value = data.getValue();
			if (value == null) {
				continue;
			}

			String text = value.toString();
			String lower = text.toLowerCase();
			ArrayList<String> matched = new ArrayList<>();
			for (String keyword : STRING_KEYWORDS) {
				if (lower.contains(keyword.toLowerCase())) {
					matched.add(keyword);
				}
			}
			if (matched.isEmpty()) {
				continue;
			}

			Reference[] refs = getReferencesTo(data.getAddress());
			for (Reference ref : refs) {
				Function function = getFunctionContaining(ref.getFromAddress());
				addFunction(interesting, function, "string:" + String.join(",", matched));
			}
		}

		FunctionIterator functionIterator = currentProgram.getFunctionManager().getFunctions(true);
		while (functionIterator.hasNext() && !monitor.isCancelled()) {
			Function function = functionIterator.next();
			String name = function.getName(true);
			String lowerName = name.toLowerCase();
			for (String keyword : FUNCTION_KEYWORDS) {
				if (lowerName.contains(keyword.toLowerCase())) {
					addFunction(interesting, function, "name:" + keyword);
				}
			}
		}

		DecompInterface decompiler = new DecompInterface();
		decompiler.openProgram(currentProgram);

		StringBuilder summary = new StringBuilder();
		summary.append("Program: ").append(currentProgram.getName()).append('\n');
		summary.append("Matched functions: ").append(interesting.size()).append('\n').append('\n');

		for (Map.Entry<String, Match> entry : interesting.entrySet()) {
			String address = entry.getKey();
			Match match = entry.getValue();
			Function function = match.function;
			String fullName = function.getName(true);
			String reasons = String.join("; ", match.reasons);

			summary.append(address).append('\t').append(fullName).append('\t').append(reasons).append('\n');

			String fileName = address.replace(':', '_') + "_" + safeName(fullName) + ".c";
			Path filePath = outputDir.resolve(fileName);

			StringBuilder content = new StringBuilder();
			content.append("FUNCTION: ").append(fullName).append('\n');
			content.append("ENTRY: ").append(address).append('\n');
			content.append("REASONS: ").append(reasons).append('\n').append('\n');

			DecompileResults result = decompiler.decompileFunction(function, 60, monitor);
			if (result != null && result.decompileCompleted() && result.getDecompiledFunction() != null) {
				content.append(result.getDecompiledFunction().getC());
			}
			else {
				String error = (result == null) ? "no result returned" : result.getErrorMessage();
				content.append("// Decompilation failed: ").append(error).append('\n');
			}

			writeText(filePath, content.toString());
		}

		writeText(outputDir.resolve("summary.txt"), summary.toString());
		println("Wrote " + interesting.size() + " decompiled functions to " + outputDir);
	}

	private void addFunction(Map<String, Match> interesting, Function function, String reason) {
		if (function == null) {
			return;
		}
		String key = function.getEntryPoint().toString();
		Match match = interesting.get(key);
		if (match == null) {
			match = new Match(function);
			interesting.put(key, match);
		}
		match.reasons.add(reason);
	}

	private String safeName(String text) {
		return text.replaceAll("[^A-Za-z0-9._-]+", "_");
	}

	private void writeText(Path path, String content) throws IOException {
		Files.write(path, content.getBytes(StandardCharsets.UTF_8));
	}
}

File: ghidra-scripts/ExtractShotPipeline.py
from ghidra.app.decompiler import DecompInterface
from ghidra.util.task import ConsoleTaskMonitor
from java.io import File
import os
import re


STRING_KEYWORDS = [
    "Wait.For.Shot",
    "waitForShot",
    "ExtractBallClub",
    "Extract.BallClub",
    "Ball.Build",
    "Club.Build",
    "BackSpinSideSpinInference",
    "Club.Detection",
    "Ball.Impact.On.Club",
    "loadCamera",
    "StartSingleFrameCamera",
    "StartBoardFrameCamera",
    "saveShot",
    "exportPostShot",
    "loadMLModel",
]

FUNCTION_KEYWORDS = [
    "waitForShot",
    "extractBallClub",
    "ballHasMoveEnough",
    "saveShot",
    "exportPostShot",
    "loadCamera",
    "loadMLModel",
    "setCameraCallback",
    "onNewShot",
    "club",
    "spin",
]


def ensure_dir(path):
    if not os.path.isdir(path):
        os.makedirs(path)


def safe_name(name):
    return re.sub(r"[^A-Za-z0-9._-]+", "_", name)


def write_text(path, content):
    fh = open(path, "w")
    try:
        fh.write(content)
    finally:
        fh.close()


args = getScriptArgs()
if len(args) < 1:
    raise Exception("Expected output directory as first argument")

output_dir = args[0]
ensure_dir(output_dir)

listing = currentProgram.getListing()
function_manager = currentProgram.getFunctionManager()
monitor = ConsoleTaskMonitor()

interesting = {}


def add_function(func, reason):
    if func is None:
        return
    key = str(func.getEntryPoint())
    if key not in interesting:
        interesting[key] = {
            "function": func,
            "reasons": []
        }
    if reason not in interesting[key]["reasons"]:
        interesting[key]["reasons"].append(reason)


data_iter = listing.getDefinedData(True)
while data_iter.hasNext() and not monitor.isCancelled():
    data = data_iter.next()
    try:
        if not data.hasStringValue():
            continue
        value = data.getValue()
        if value is None:
            continue
        text = str(value)
    except:
        continue

    matched = []
    lower_text = text.lower()
    for keyword in STRING_KEYWORDS:
        if keyword.lower() in lower_text:
            matched.append(keyword)

    if not matched:
        continue

    refs = getReferencesTo(data.getAddress())
    for ref in refs:
        func = getFunctionContaining(ref.getFromAddress())
        if func is None:
            continue
        add_function(func, "string:" + ",".join(matched))


func_iter = function_manager.getFunctions(True)
while func_iter.hasNext() and not monitor.isCancelled():
    func = func_iter.next()
    try:
        name = func.getName(True)
    except:
        name = func.getName()
    lower_name = name.lower()
    for keyword in FUNCTION_KEYWORDS:
        if keyword.lower() in lower_name:
            add_function(func, "name:" + keyword)


decompiler = DecompInterface()
decompiler.openProgram(currentProgram)

summary_lines = []
summary_lines.append("Program: %s" % currentProgram.getName())
summary_lines.append("Matched functions: %d" % len(interesting))
summary_lines.append("")

for entry in sorted(interesting.keys()):
    func = interesting[entry]["function"]
    reasons = "; ".join(sorted(interesting[entry]["reasons"]))
    try:
        full_name = func.getName(True)
    except:
        full_name = func.getName()

    summary_lines.append("%s\t%s\t%s" % (entry, full_name, reasons))

    file_basename = "%s_%s" % (entry.replace(":", "_"), safe_name(full_name))
    result = decompiler.decompileFunction(func, 60, monitor)

    content_lines = []
    content_lines.append("FUNCTION: %s" % full_name)
    content_lines.append("ENTRY: %s" % entry)
    content_lines.append("REASONS: %s" % reasons)
    content_lines.append("")

    if result is not None and result.decompileCompleted():
        decompiled = result.getDecompiledFunction()
        if decompiled is not None:
            content_lines.append(decompiled.getC())
        else:
            content_lines.append("// Decompilation produced no function body.")
    else:
        message = ""
        try:
            message = result.getErrorMessage()
        except:
            message = "unknown error"
        content_lines.append("// Decompilation failed: %s" % message)

    write_text(os.path.join(output_dir, file_basename + ".c"), "\n".join(content_lines))


write_text(os.path.join(output_dir, "summary.txt"), "\n".join(summary_lines))
print("Wrote %d decompiled functions to %s" % (len(interesting), output_dir))
